### ISSUE TITLE
Rule S6265: Granting access to S3 buckets to all or authenticated users is security-sensitive

### ISSUE THREAD

**Issue by pedro-oliveira-sonarsource** [SonarSource/SonarJS#3198]
## WHAT
The rule detects when an S3 bucket is configured to be accessible publicly using the JavaScript AWS CDK  library. It does not cover custom policy statement definitions. This has been previously implemented for Python, see [SONARPY-1011](https://github.com/SonarSource/sonar-python/pull/1129).

Specification ticket: [APPSEC-39](https://jira.sonarsource.com/browse/APPSEC-39)
[MMF-2812](https://jira.sonarsource.com/browse/MMF-2812)

## HOW
Detection logic

The rule should detect the following patterns:

1. When `Bucket` constructor is called with an improper configuration
2. When `Bucket.grantPublicAccess` is called
3. When `BucketDeployment` constructor is called with an improper configuration

### `Bucket` constructor

- Method: `aws-cdk-lib.aws_s3.Bucket` constructor
- Argument: `props` 3rd positional argument
- Sensitive properties: 
   - `accessControl`
      - Sensitive values: `PUBLIC_READ`, `PUBLIC_READ_WRITE`, `AUTHENTICATED_READ`
   - `publicReadAccess`
      - Sensitive value: `true` 

The rule raises an issue when a`Bucket` is constructed with one of the above properties set to the incorrect value.
```javascript
new s3.Bucket(this, 'id', {
    bucketName: 'bucketnoncompliant',
    accessControl: s3.BucketAccessControl.PUBLIC_READ_WRITE // Sensitive
);

const publicReadAccess = true; // Secondary Location
new s3.Bucket(this, 'id', {
    bucketName: 'bucketnoncompliant',
    publicReadAccess: publicReadAccess // Sensitive
);
```

### `Bucket.grantPublicAccess` method

Method: `aws-cdk-lib.aws_s3.Bucket.grantPublicAccess`

Rule raises on the Bucket construct class when _grantPublicAccess_ method is called.
```javascript
const bucketNoncompliant = new s3.Bucket(this, 'id', {
    bucketName: 'bucketnoncompliant'
);

bucketNoncompliant.grantPublicAccess(); // Sensitive
```

### `BucketDeployment` constructor

- Method: `aws-cdk-lib.aws_s3.BucketDeployment` constructor
- Argument: `props` 3rd positional argument
- Sensitive propertie:`accessControl`
   - Sensitive values: `PUBLIC_READ`, `PUBLIC_READ_WRITE`, `AUTHENTICATED_READ`

The rule raises an issue when a `BucketDeployment` is contructed with the above propertie set to the incorrect value.
```javascript
const bucketNoncompliant = new s3.Bucket(this, 'id', {
    bucketName: 'bucketnoncompliant'
);

new s3deploy.BucketDeployment(this, 'upload', {
    accessControl: s3.BucketAccessControl.PUBLIC_READ_WRITE // Sensitive
});
```

## Code examples

https://github.com/SonarSource/security-expected-issues/pull/721

## Message and highlighting

See https://github.com/SonarSource/rspec/pull/1084

**Comment by pierre-loup-tristant-sonarsource** [comment]
Reading again the title I think I think it can be improved. You suggested, "Granting public access to S3 buckets is security-sensitive". Let's go with that.
